pkg/cache/snapshot.go

```diff
@@ -0,0 +1,297 @@
+// Copyright 2018 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/envoyproxy/go-control-plane/pkg/log"
+)
+
+// SnapshotCache is a snapshot-based cache that maintains a single versioned
+// snapshot of responses per node. SnapshotCache consistently replies with the
+// latest snapshot. For the protocol to work correctly, EDS/RDS requests are
+// responded only when all resources in the snapshot xDS response are named as
+// part of the request. It is expected that the CDS response names all EDS
+// clusters, and the LDS response names all RDS routes in a snapshot, to ensure
+// that Envoy makes the request for all EDS clusters or RDS routes eventually.
+type SnapshotCache struct {
+	log log.Logger
+
+	// snapshots are cached resources
+	snapshots map[string]Snapshot
+
+	// status information for all nodes
+	status map[string]StatusInfo
+
+	// callback requests missing responses
+	callback func(string)
```

Comment:
is this used anymore? It seems like you need it in `CreateWatch` to tell whatever the snapshot provider is that you need a snapshot for a given key.

(also, I think you need some way to signal to the snapshot provider that you no longer need the key, or it will keep providing snapshots forever)

Reply:
ah, I see. thanks.